### ISSUE TITLE
pyright multi-assignment tuple type propagation might be broken

### ISSUE THREAD

**Issue by shoffmeister** [microsoft/pylance-release#1923]
Issue Type: <b>Bug</b>

Python allows having more than one assignment on a single line (https://docs.python.org/3/reference/simple_stmts.html#assignment-statements). If tuples are involved, it seems as if type propagation takes place that is not wanted.

Given
```python
SOME_BUFFER: bytes = b"abc"

def takeme(value: tuple[int, int]) -> None: ...

v1, v2 = tuple(SOME_BUFFER[0:2])
takeme((v1, v2))
v3 = (v1, v2)
takeme(v3)

x3 = x1, x2 = tuple(SOME_BUFFER[0:2])
takeme(x3)

y1, y2 = y3 = tuple(SOME_BUFFER[0:2])
takeme(y3)
```
pyright (and mypy) will complain that `Argument of type "tuple[int, ...]" cannot be assigned to parameter "value" of type "tuple[int, int]" in function "takeme" - Tuple size mismatch; expected 2 but received indeterminate number`

a) It's a bit unfortunate that `tuple(SOME_BUFFER[0:2])` is not computed as `tuple[int, int]` but as `tuple[int, ...]`; after all, `2` is a constant. If the type inference engine was able to evaluate constants, this issue would never have been raised ;)

b) Given that we have `tuple[int, ...]` on the RHS, I am uncertain about how assignment actually is supposed to work in Python; there are two possible long-form pseudo-representations:

b1) repeated assignment of RHS
```python
my_temp = tuple(SOME_BUFFER[0:2])
x1, x2 = my_temp
x3 = my_temp
```
b2) chained assignment
```python
my_temp = tuple(SOME_BUFFER[0:2])
x1, x2 = my_temp
x3 = (x1, x2)
```
From the diagnostics on the sample code it seems as if "repeated assignment of RHS" is the chosen algorithm. I do wonder, though, whether this is in alignment with the documentation at https://docs.python.org/3/reference/simple_stmts.html#assignment-statements:

>An assignment statement evaluates the expression list (remember that this can be a single expression or a comma-separated list, the latter yielding a tuple) and assigns the single resulting object to each of the target lists, from left to right.
>Assignment is defined recursively depending on the form of the target (list). 

My interpretation of the Python documentation is that in my sample code, one of `takeme(x3)` and `takeme(y3)` should succeed (currently both result in error diagnostics), because, effectively, "defined recursively" implies "chained assignment"?

But perhaps my reading of the Python documentation is inadequate?

*****
Extension version: 2021.10.0
VS Code version: Code 1.61.0 (ee8c7def80afc00dd6e593ef12f37756d8f504ea, 2021-10-07T18:13:09.652Z)
OS version: Windows_NT x64 10.0.18363
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz (8 x 2904)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.66GB (44.04GB free)|
|Process Argv|--crash-reporter-id 5ddcd13a-5853-4d88-9c59-31e3be66729a|
|Screen Reader|no|
|VM|44%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30300192
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30358481
pythondataviewer:30285071
pythonvsuse255:30340121
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
vsccppwtct:30378365
pythonvssor306:30344512
bridge0708:30335490
pygetstartedt2:30371810
dockerwalkthru:30377721
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedc:30364665
pythonvspyt187:30373474
pydsgst2:30361792
vsqsis200:30374795
vsaa593cf:30376535
vssld246cf:30379912

```

</details>

**Comment by erictraut** [comment]
I think that both of your representations are correct and are equivalent.

The type of the expression `tuple(SOME_BUFFER[0:2])` is evaluated as `tuple[int, ...]`. That's a straightforward application of the type rules. The type of subexpression `SOME_BUFFER[0:2]` is `bytes`. The type declaration for the `tuple` constructor is as follows:

```python
class tuple(Sequence[_T_co], Generic[_T_co]):
    def __new__(cls: Type[_T], iterable: Iterable[_T_co] = ...) -> _T: ...
```

Since `bytes` derives from `Sequence[int]`, the call signature for the constructor is satisfied, and the resulting type is `tuple[int, ...]`.

One could argue that the type of the RHS of the assignment (`tuple[int, ...]`) could be implicitly narrowed to `tuple[int, int]` by its assignment to `x1, x2` using the logic that if the assigned tuple had any length other than 2, this assignment would result in a runtime exception. That's a pretty sophisticated form of type narrowing though. It's currently not supported by mypy or pyright (or any other type checker that I'm aware of). It would be quite different from other forms of type narrowing in pyright, and therefore a pretty heavy lift to add this. We might consider doing this if it were a pattern that arises frequently in Python code, but I don't currently have evidence that it is. Its use would be very specific to tuple assignments that involve destructuring (unpacking) where the RHS is a tuple of unknown length and the LHS is a known length (i.e. it doesn't involve any variables with `*` operators) and where the result is immediately (within the same statement) assigned to another variable. That is, it wouldn't generalize to cases other than the very narrow case in your example.

In summary, the current behavior is as expected, and I don't think there's a bug. There's a potential feature request, but I don't think we have enough evidence at this time that its cost and complexity is merited. I'm going to close the issue for now, but we can reconsider in the future if there is sufficient interest.

**Comment by shoffmeister** [comment]
Thank you for the extensive response on point a) - I agree, I had missed that SOME_BUFFER[0..2] returns _just_ bytes, which has no *dimension* only a *size* (length).

Would you also be able to comment on point b)? b) is the result of agreeing to a) being correct. I could also open a different issue for that, might be better, more structured?
